## Library Import
For this task Turicreate package from Apple.Inc was used and this library only runs on MacOs or on a Linux Kernel

In [51]:
import turicreate as tc 
import pandas as pd
import numpy as np

## Data Import

the Id of the customer, the id of the product and the review score were used for this task

In [52]:
cols=["customer_unique_id", "product_id", "review_score"]
df= pd.read_csv("python_merged.csv", usecols=cols) #use only columns in cols

## Data cleaning 
Duplicates and Null values were eliminated

In [53]:
df.dropna(inplace=True) 
df.drop_duplicates(inplace=True, keep="first")

# Data Manipulation
The data are both normalized and standardized

In [54]:
df_norm= df.copy()
  
# apply normalization techniques by Column 1
column = 'review_score'
df_norm[column] = (df_norm[column] - df_norm[column].min()) / (df_norm[column].max() - df_norm[column].min())
df_norm[column]=(df_norm[column]-df_norm[column].mean())/df[column].std()
  
# view normalized data
display(df_norm)

product_id  review_score  \
0       f3c2d01a84c947b078e32bbef0718962     -0.387068   
1       5a6b04657a4c5ee34285d1e4619a96b4     -0.015951   
4       e2a1d45a73dc7f5a7f9236b043431b89      0.169608   
5       817e1c2d22418c36386406ccacfa53e8      0.169608   
6       9e93b2c4cb5eea05e75a481c129b104d     -0.572627   
...                                  ...           ...   
118138  17e18b0c88a853dd6de3e48a7cfa9d9a     -0.572627   
118139  6aa063e063f2ab982b471e58afe06d72     -0.387068   
118140  282b126b2354516c5f400154398f616d      0.169608   
118141  96ea060e41bdecc64e2de00b97068975      0.169608   
118142  87b92e06b320e803d334ac23966c80b1     -0.572627   

                      customer_unique_id  
0       4854e9b3feff728c13ee5fc7d1547e92  
1       830d5b7aaa3b6f1e9ad63703bec97d23  
4       87776adb449c551e74c13fc34f036105  
5       9f302d00dd3e18ed3745778184b4f0fe  
6       3f4f614c632af7fc7508462a7cb55ac2  
...                                  ...  
118138  56b6eede1b10925212f054a7ba614796  
118139  101375bf617fd60c9eee42f98d9a73d6  
118140  b030929cf3b8c3370ea8c611f9ccb32e  
118141  3977f83a14549e6265bcded84e92ee80  
118142  44b6bbfea26596437062a38c8e6bcec1  

[100638 rows x 3 columns]

## Model Preparation

The dataframe is first split into train and test with a 80/20 proportion, then they are transformed into an SFrame which is a specific formt of the package Turicreate, this has the advantage of not relying on RAM but on physical memory

In [55]:
def split_data(data):
    '''
    Splits dataset into training and test set.
    
    Args:
        data (pandas.DataFrame)
        
    Returns
        train_data (tc.SFrame)
        test_data (tc.SFrame)
    '''
    train, test = train_test_split(data, test_size = .2)
    train_data = tc.SFrame(train)
    test_data = tc.SFrame(test)
    return train_data, test_data

In [56]:
from sklearn.model_selection import train_test_split
train_data, test_data = split_data(df) #the function is called
train_data_norm, test_data_norm = split_data(df_norm)

## Modelling

First, the number of product to recommend is set to 5

In [57]:
user_id = 'customer_unique_id'
item_id = 'product_id'
users_to_recommend = list(df[user_id])
n_rec = 5 # number of items to recommend
n_display = 30 # to display the first few rows in an output dataset

Then a function is prepared to include the measures which are selected to perform the recommendation 
Three methods have been selected:
**Popularity method**: which is based on user preferences
**Cosine Similarity**: a collaborative filtering method
**Pearson Correlation**: another collaborative filtering method

In [58]:
def model(train_data, name, user_id, item_id, target, users_to_recommend, n_rec, n_display):
    if name == 'popularity':
        model = tc.popularity_recommender.create(train_data, 
                                                    user_id=user_id, 
                                                    item_id=item_id, 
                                                    target=target)
    elif name == 'cosine':
        model = tc.item_similarity_recommender.create(train_data, 
                                                    user_id=user_id, 
                                                    item_id=item_id, 
                                                    target=target, 
                                                    similarity_type='cosine')
    elif name == 'pearson':
        model = tc.item_similarity_recommender.create(train_data, 
                                                    user_id=user_id, 
                                                    item_id=item_id, 
                                                    target=target, 
                                                    similarity_type='pearson')
        
    recom = model.recommend(users=users_to_recommend, k=n_rec)
    recom.print_rows(n_display)
    return model

## Run models on train data

The output here shows what has been recommended to each customer, each customer has 5 products recommended to them ranked from the first to the fifth. 
Here 6 outputs are produced: three for the non normalized dataset and three for the transformed dataset

In [59]:
name = 'popularity'
target = 'review_score'
popularity = model(train_data, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 80510 observations with 76129 users and 28332 items.

Data prepared in: 0.110956s

80510 observations to process; with 28332 unique items.

recommendations finished on 1000/100638 queries. users per second: 21262.6

recommendations finished on 2000/100638 queries. users per second: 21608.1

recommendations finished on 3000/100638 queries. users per second: 20655.5

recommendations finished on 4000/100638 queries. users per second: 20439.8

recommendations finished on 5000/100638 queries. users per second: 20549.1

recommendations finished on 6000/100638 queries. users per second: 20448.2

recommendations finished on 7000/100638 queries. users per second: 20273.9

recommendations finished on 8000/100638 queries. users per second: 20304.4

recommendations finished on 9000/100638 queries. users per second: 20357.8

recommendations finished on 10000/100638 queries. users per second: 20385.1

recommendations finished on 11000/100638 queries. users per second: 20435.8

recommendations finished on 12000/100638 queries. users per second: 20468.1

recommendations finished on 13000/100638 queries. users per second: 20535.9

recommendations finished on 14000/100638 queries. users per second: 20565.9

recommendations finished on 15000/100638 queries. users per second: 20551.2

recommendations finished on 16000/100638 queries. users per second: 20541.4

recommendations finished on 17000/100638 queries. users per second: 20538.9

recommendations finished on 18000/100638 queries. users per second: 20529.2

recommendations finished on 19000/100638 queries. users per second: 20408.5

recommendations finished on 20000/100638 queries. users per second: 20422.4

recommendations finished on 21000/100638 queries. users per second: 20427.3

recommendations finished on 22000/100638 queries. users per second: 20392

recommendations finished on 23000/100638 queries. users per second: 20405.8

recommendations finished on 24000/100638 queries. users per second: 20420.9

recommendations finished on 25000/100638 queries. users per second: 20439

recommendations finished on 26000/100638 queries. users per second: 20461.6

recommendations finished on 27000/100638 queries. users per second: 20495.8

recommendations finished on 28000/100638 queries. users per second: 20500.8

recommendations finished on 29000/100638 queries. users per second: 20508.2

recommendations finished on 30000/100638 queries. users per second: 20504.5

recommendations finished on 31000/100638 queries. users per second: 20481.6

recommendations finished on 32000/100638 queries. users per second: 20492.6

recommendations finished on 33000/100638 queries. users per second: 20506.7

recommendations finished on 34000/100638 queries. users per second: 20506.6

recommendations finished on 35000/100638 queries. users per second: 20501

recommendations finished on 36000/100638 queries. users per second: 20495.3

recommendations finished on 37000/100638 queries. users per second: 20492

recommendations finished on 38000/100638 queries. users per second: 20499.8

recommendations finished on 39000/100638 queries. users per second: 20486.1

recommendations finished on 40000/100638 queries. users per second: 20483.7

recommendations finished on 41000/100638 queries. users per second: 20474.3

recommendations finished on 42000/100638 queries. users per second: 20493.2

recommendations finished on 43000/100638 queries. users per second: 20420

recommendations finished on 44000/100638 queries. users per second: 20424.7

recommendations finished on 45000/100638 queries. users per second: 20423.4

recommendations finished on 46000/100638 queries. users per second: 20403.2

recommendations finished on 47000/100638 queries. users per second: 20401.1

recommendations finished on 48000/100638 queries. users per second: 20405.4

recommendations finished on 49000/100638 queries. users per second: 20417.8

recommendations finished on 50000/100638 queries. users per second: 20428.2

recommendations finished on 51000/100638 queries. users per second: 20438.7

recommendations finished on 52000/100638 queries. users per second: 20405.5

recommendations finished on 53000/100638 queries. users per second: 20432

recommendations finished on 54000/100638 queries. users per second: 20475.2

recommendations finished on 55000/100638 queries. users per second: 20481.5

recommendations finished on 56000/100638 queries. users per second: 20425

recommendations finished on 57000/100638 queries. users per second: 20427.8

recommendations finished on 58000/100638 queries. users per second: 20419.9

recommendations finished on 59000/100638 queries. users per second: 20449.8

recommendations finished on 60000/100638 queries. users per second: 20434.2

recommendations finished on 61000/100638 queries. users per second: 20444.6

recommendations finished on 62000/100638 queries. users per second: 20473.6

recommendations finished on 63000/100638 queries. users per second: 20458.2

recommendations finished on 64000/100638 queries. users per second: 20460.1

recommendations finished on 65000/100638 queries. users per second: 20452.6

recommendations finished on 66000/100638 queries. users per second: 20477.5

recommendations finished on 67000/100638 queries. users per second: 20459.8

recommendations finished on 68000/100638 queries. users per second: 20474.3

recommendations finished on 69000/100638 queries. users per second: 20451.7

recommendations finished on 70000/100638 queries. users per second: 20479.1

recommendations finished on 71000/100638 queries. users per second: 20498.7

recommendations finished on 72000/100638 queries. users per second: 20492.4

recommendations finished on 73000/100638 queries. users per second: 20516.8

recommendations finished on 74000/100638 queries. users per second: 20540.3

recommendations finished on 75000/100638 queries. users per second: 20568.6

recommendations finished on 76000/100638 queries. users per second: 20584.4

recommendations finished on 77000/100638 queries. users per second: 20546

recommendations finished on 78000/100638 queries. users per second: 20569.6

recommendations finished on 79000/100638 queries. users per second: 20552.6

recommendations finished on 80000/100638 queries. users per second: 20569.8

recommendations finished on 81000/100638 queries. users per second: 20590.8

recommendations finished on 82000/100638 queries. users per second: 20609.4

recommendations finished on 83000/100638 queries. users per second: 20627.2

recommendations finished on 84000/100638 queries. users per second: 20646.2

recommendations finished on 85000/100638 queries. users per second: 20658.9

recommendations finished on 86000/100638 queries. users per second: 20624.9

recommendations finished on 87000/100638 queries. users per second: 20634.2

recommendations finished on 88000/100638 queries. users per second: 20652.8

recommendations finished on 89000/100638 queries. users per second: 20670.7

recommendations finished on 90000/100638 queries. users per second: 20686.9

recommendations finished on 91000/100638 queries. users per second: 20702.3

recommendations finished on 92000/100638 queries. users per second: 20720

recommendations finished on 93000/100638 queries. users per second: 20733.8

recommendations finished on 94000/100638 queries. users per second: 20739.9

recommendations finished on 95000/100638 queries. users per second: 20746.7

recommendations finished on 96000/100638 queries. users per second: 20749.8

recommendations finished on 97000/100638 queries. users per second: 20750.4

recommendations finished on 98000/100638 queries. users per second: 20744.3

recommendations finished on 99000/100638 queries. users per second: 20694.4

recommendations finished on 100000/100638 queries. users per second: 20480.8

+-------------------------------+-------------------------------+-------+------+
|       customer_unique_id      |           product_id          | score | rank |
+-------------------------------+-------------------------------+-------+------+
| 4854e9b3feff728c13ee5fc7d1... | 6ed93af03d1f53308d3a9c6555... |  5.0  |  1   |
| 4854e9b3feff728c13ee5fc7d1... | f4851ffef385d506603396041d... |  5.0  |  2   |
| 4854e9b3feff728c13ee5fc7d1... | 18047039004204f52054f74b95... |  5.0  |  3   |
| 4854e9b3feff728c13ee5fc7d1... | 97868e08c02c9f4f6b49ebe5a4... |  5.0  |  4   |
| 4854e9b3feff728c13ee5fc7d1... | 73abf6f6dfa98d1d04760f4cc1... |  5.0  |  5   |
| 830d5b7aaa3b6f1e9ad63703be... | 6ed93af03d1f53308d3a9c6555... |  5.0  |  1   |
| 830d5b7aaa3b6f1e9ad63703be... | f4851ffef385d506603396041d... |  5.0  |  2   |
| 830d5b7aaa3b6f1e9ad63703be... | 18047039004204f52054f74b95... |  5.0  |  3   |
| 830d5b7aaa3b6f1e9ad63703be... | 97868e08c02c9f4f6b49ebe5a4... |  5.0  |  4   |
| 830d5b7aaa3b6f1e9ad63703be

In [60]:
name = 'popularity'
target = 'review_score'
pop_norm = model(train_data_norm, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 80510 observations with 76144 users and 28388 items.

Data prepared in: 0.129251s

80510 observations to process; with 28388 unique items.

recommendations finished on 1000/100638 queries. users per second: 19220.8

recommendations finished on 2000/100638 queries. users per second: 19892.2

recommendations finished on 3000/100638 queries. users per second: 20223.3

recommendations finished on 4000/100638 queries. users per second: 19935.8

recommendations finished on 5000/100638 queries. users per second: 20044.3

recommendations finished on 6000/100638 queries. users per second: 20048.7

recommendations finished on 7000/100638 queries. users per second: 20122.1

recommendations finished on 8000/100638 queries. users per second: 20217.1

recommendations finished on 9000/100638 queries. users per second: 20309.8

recommendations finished on 10000/100638 queries. users per second: 20472

recommendations finished on 11000/100638 queries. users per second: 20467.1

recommendations finished on 12000/100638 queries. users per second: 20399.1

recommendations finished on 13000/100638 queries. users per second: 20472.3

recommendations finished on 14000/100638 queries. users per second: 20464.5

recommendations finished on 15000/100638 queries. users per second: 20494.2

recommendations finished on 16000/100638 queries. users per second: 20484.1

recommendations finished on 17000/100638 queries. users per second: 20466.1

recommendations finished on 18000/100638 queries. users per second: 20451.2

recommendations finished on 19000/100638 queries. users per second: 20366.3

recommendations finished on 20000/100638 queries. users per second: 20355.1

recommendations finished on 21000/100638 queries. users per second: 20312.7

recommendations finished on 22000/100638 queries. users per second: 20221.8

recommendations finished on 23000/100638 queries. users per second: 20266.8

recommendations finished on 24000/100638 queries. users per second: 20219.3

recommendations finished on 25000/100638 queries. users per second: 20214.9

recommendations finished on 26000/100638 queries. users per second: 20259.4

recommendations finished on 27000/100638 queries. users per second: 20234.1

recommendations finished on 28000/100638 queries. users per second: 20228

recommendations finished on 29000/100638 queries. users per second: 20216.2

recommendations finished on 30000/100638 queries. users per second: 20213.8

recommendations finished on 31000/100638 queries. users per second: 20227.2

recommendations finished on 32000/100638 queries. users per second: 20232.6

recommendations finished on 33000/100638 queries. users per second: 20256.3

recommendations finished on 34000/100638 queries. users per second: 20249.5

recommendations finished on 35000/100638 queries. users per second: 20261.3

recommendations finished on 36000/100638 queries. users per second: 20308.5

recommendations finished on 37000/100638 queries. users per second: 20295.1

recommendations finished on 38000/100638 queries. users per second: 20305.5

recommendations finished on 39000/100638 queries. users per second: 20317.5

recommendations finished on 40000/100638 queries. users per second: 20351.6

recommendations finished on 41000/100638 queries. users per second: 20317.6

recommendations finished on 42000/100638 queries. users per second: 20323.3

recommendations finished on 43000/100638 queries. users per second: 20346.9

recommendations finished on 44000/100638 queries. users per second: 20342.5

recommendations finished on 45000/100638 queries. users per second: 20339.9

recommendations finished on 46000/100638 queries. users per second: 20336.1

recommendations finished on 47000/100638 queries. users per second: 20350.1

recommendations finished on 48000/100638 queries. users per second: 20353.2

recommendations finished on 49000/100638 queries. users per second: 20339.9

recommendations finished on 50000/100638 queries. users per second: 20257.9

recommendations finished on 51000/100638 queries. users per second: 20255.6

recommendations finished on 52000/100638 queries. users per second: 20217.3

recommendations finished on 53000/100638 queries. users per second: 20257.6

recommendations finished on 54000/100638 queries. users per second: 20215.1

recommendations finished on 55000/100638 queries. users per second: 20257

recommendations finished on 56000/100638 queries. users per second: 20226.2

recommendations finished on 57000/100638 queries. users per second: 20190.4

recommendations finished on 58000/100638 queries. users per second: 20200.9

recommendations finished on 59000/100638 queries. users per second: 20185.2

recommendations finished on 60000/100638 queries. users per second: 20160.3

recommendations finished on 61000/100638 queries. users per second: 20153.7

recommendations finished on 62000/100638 queries. users per second: 20179.5

recommendations finished on 63000/100638 queries. users per second: 20162

recommendations finished on 64000/100638 queries. users per second: 20192.6

recommendations finished on 65000/100638 queries. users per second: 20177.5

recommendations finished on 66000/100638 queries. users per second: 20206.5

recommendations finished on 67000/100638 queries. users per second: 20221.6

recommendations finished on 68000/100638 queries. users per second: 20221.8

recommendations finished on 69000/100638 queries. users per second: 20244.2

recommendations finished on 70000/100638 queries. users per second: 20222.4

recommendations finished on 71000/100638 queries. users per second: 20243.5

recommendations finished on 72000/100638 queries. users per second: 20225.9

recommendations finished on 73000/100638 queries. users per second: 20250.7

recommendations finished on 74000/100638 queries. users per second: 20223

recommendations finished on 75000/100638 queries. users per second: 20235.7

recommendations finished on 76000/100638 queries. users per second: 20256.4

recommendations finished on 77000/100638 queries. users per second: 20232.3

recommendations finished on 78000/100638 queries. users per second: 20256.6

recommendations finished on 79000/100638 queries. users per second: 20275.6

recommendations finished on 80000/100638 queries. users per second: 20293

recommendations finished on 81000/100638 queries. users per second: 20316.1

recommendations finished on 82000/100638 queries. users per second: 20334.8

recommendations finished on 83000/100638 queries. users per second: 20351.8

recommendations finished on 84000/100638 queries. users per second: 20373.7

recommendations finished on 85000/100638 queries. users per second: 20393.6

recommendations finished on 86000/100638 queries. users per second: 20416.5

recommendations finished on 87000/100638 queries. users per second: 20436.9

recommendations finished on 88000/100638 queries. users per second: 20456

recommendations finished on 89000/100638 queries. users per second: 20456

recommendations finished on 90000/100638 queries. users per second: 20469.6

recommendations finished on 91000/100638 queries. users per second: 20404.6

recommendations finished on 92000/100638 queries. users per second: 20412

recommendations finished on 93000/100638 queries. users per second: 20404.4

recommendations finished on 94000/100638 queries. users per second: 20421.2

recommendations finished on 95000/100638 queries. users per second: 20400.5

recommendations finished on 96000/100638 queries. users per second: 20410.9

recommendations finished on 97000/100638 queries. users per second: 20416.1

recommendations finished on 98000/100638 queries. users per second: 20399

recommendations finished on 99000/100638 queries. users per second: 20360.4

recommendations finished on 100000/100638 queries. users per second: 20253.2

+-------------------------------+-------------------------------+
|       customer_unique_id      |           product_id          |
+-------------------------------+-------------------------------+
| 4854e9b3feff728c13ee5fc7d1... | 6c50c87e36d8641d67fbcceed5... |
| 4854e9b3feff728c13ee5fc7d1... | d0349534ab9b46b98d13a53051... |
| 4854e9b3feff728c13ee5fc7d1... | f196248e8b5d060cca414a664e... |
| 4854e9b3feff728c13ee5fc7d1... | c0350d6ac413eda4641bf92ab6... |
| 4854e9b3feff728c13ee5fc7d1... | 3215010238fcd9cab6ba7d2b81... |
| 830d5b7aaa3b6f1e9ad63703be... | 6c50c87e36d8641d67fbcceed5... |
| 830d5b7aaa3b6f1e9ad63703be... | d0349534ab9b46b98d13a53051... |
| 830d5b7aaa3b6f1e9ad63703be... | f196248e8b5d060cca414a664e... |
| 830d5b7aaa3b6f1e9ad63703be... | c0350d6ac413eda4641bf92ab6... |
| 830d5b7aaa3b6f1e9ad63703be... | 3215010238fcd9cab6ba7d2b81... |
| 87776adb449c551e74c13fc34f... | 6c50c87e36d8641d67fbcceed5... |
| 87776adb449c551e74c13fc34f... | d0349534ab9b46b98d13a53051... |
| 87776adb

In [61]:
name="cosine"
target="review_score"
cosine= model(train_data, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 80510 observations with 76129 users and 28332 items.

Data prepared in: 0.128567s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.592ms                        | 1.25       |

| 25.359ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.52s                               | 0                | 0               |

| 3.78s                               | 100              | 28332           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 3.80741s

recommendations finished on 1000/100638 queries. users per second: 14857.7

recommendations finished on 2000/100638 queries. users per second: 14017.4

recommendations finished on 3000/100638 queries. users per second: 13878.4

recommendations finished on 4000/100638 queries. users per second: 13963.7

recommendations finished on 5000/100638 queries. users per second: 13963.6

recommendations finished on 6000/100638 queries. users per second: 13814.5

recommendations finished on 7000/100638 queries. users per second: 13551.1

recommendations finished on 8000/100638 queries. users per second: 13540.3

recommendations finished on 9000/100638 queries. users per second: 13579.9

recommendations finished on 10000/100638 queries. users per second: 13639.5

recommendations finished on 11000/100638 queries. users per second: 13506.7

recommendations finished on 12000/100638 queries. users per second: 13491.8

recommendations finished on 13000/100638 queries. users per second: 13481.4

recommendations finished on 14000/100638 queries. users per second: 13379.6

recommendations finished on 15000/100638 queries. users per second: 13300.1

recommendations finished on 16000/100638 queries. users per second: 13209.2

recommendations finished on 17000/100638 queries. users per second: 13148.2

recommendations finished on 18000/100638 queries. users per second: 13141.9

recommendations finished on 19000/100638 queries. users per second: 13150.8

recommendations finished on 20000/100638 queries. users per second: 13157.5

recommendations finished on 21000/100638 queries. users per second: 13137.7

recommendations finished on 22000/100638 queries. users per second: 13097.3

recommendations finished on 23000/100638 queries. users per second: 13133.3

recommendations finished on 24000/100638 queries. users per second: 13128.3

recommendations finished on 25000/100638 queries. users per second: 13108.1

recommendations finished on 26000/100638 queries. users per second: 13153.6

recommendations finished on 27000/100638 queries. users per second: 13179

recommendations finished on 28000/100638 queries. users per second: 13205.7

recommendations finished on 29000/100638 queries. users per second: 13238.7

recommendations finished on 30000/100638 queries. users per second: 13240

recommendations finished on 31000/100638 queries. users per second: 13236.2

recommendations finished on 32000/100638 queries. users per second: 13232.3

recommendations finished on 33000/100638 queries. users per second: 13205

recommendations finished on 34000/100638 queries. users per second: 13183.2

recommendations finished on 35000/100638 queries. users per second: 13229.4

recommendations finished on 36000/100638 queries. users per second: 13248.9

recommendations finished on 37000/100638 queries. users per second: 13226.9

recommendations finished on 38000/100638 queries. users per second: 13217.9

recommendations finished on 39000/100638 queries. users per second: 13207.2

recommendations finished on 40000/100638 queries. users per second: 13235

recommendations finished on 41000/100638 queries. users per second: 13227.9

recommendations finished on 42000/100638 queries. users per second: 13235.6

recommendations finished on 43000/100638 queries. users per second: 13260.4

recommendations finished on 44000/100638 queries. users per second: 13244.2

recommendations finished on 45000/100638 queries. users per second: 13236.1

recommendations finished on 46000/100638 queries. users per second: 13243

recommendations finished on 47000/100638 queries. users per second: 13233.1

recommendations finished on 48000/100638 queries. users per second: 13242.1

recommendations finished on 49000/100638 queries. users per second: 13261.6

recommendations finished on 50000/100638 queries. users per second: 13258.9

recommendations finished on 51000/100638 queries. users per second: 13271.4

recommendations finished on 52000/100638 queries. users per second: 13288.3

recommendations finished on 53000/100638 queries. users per second: 13304.2

recommendations finished on 54000/100638 queries. users per second: 13306.3

recommendations finished on 55000/100638 queries. users per second: 13315.9

recommendations finished on 56000/100638 queries. users per second: 13321.8

recommendations finished on 57000/100638 queries. users per second: 13348.8

recommendations finished on 58000/100638 queries. users per second: 13363.4

recommendations finished on 59000/100638 queries. users per second: 13376.3

recommendations finished on 60000/100638 queries. users per second: 13380

recommendations finished on 61000/100638 queries. users per second: 13392.3

recommendations finished on 62000/100638 queries. users per second: 13404.9

recommendations finished on 63000/100638 queries. users per second: 13420.5

recommendations finished on 64000/100638 queries. users per second: 13436.5

recommendations finished on 65000/100638 queries. users per second: 13452.4

recommendations finished on 66000/100638 queries. users per second: 13467.9

recommendations finished on 67000/100638 queries. users per second: 13482.6

recommendations finished on 68000/100638 queries. users per second: 13494.9

recommendations finished on 69000/100638 queries. users per second: 13505.1

recommendations finished on 70000/100638 queries. users per second: 13518.6

recommendations finished on 71000/100638 queries. users per second: 13527.4

recommendations finished on 72000/100638 queries. users per second: 13540.7

recommendations finished on 73000/100638 queries. users per second: 13552.7

recommendations finished on 74000/100638 queries. users per second: 13559.9

recommendations finished on 75000/100638 queries. users per second: 13570.4

recommendations finished on 76000/100638 queries. users per second: 13578.1

recommendations finished on 77000/100638 queries. users per second: 13582.1

recommendations finished on 78000/100638 queries. users per second: 13588.8

recommendations finished on 79000/100638 queries. users per second: 13599.5

recommendations finished on 80000/100638 queries. users per second: 13608.9

recommendations finished on 81000/100638 queries. users per second: 13621.1

recommendations finished on 82000/100638 queries. users per second: 13631.2

recommendations finished on 83000/100638 queries. users per second: 13638.1

recommendations finished on 84000/100638 queries. users per second: 13644

recommendations finished on 85000/100638 queries. users per second: 13652.9

recommendations finished on 86000/100638 queries. users per second: 13656.5

recommendations finished on 87000/100638 queries. users per second: 13664.5

recommendations finished on 88000/100638 queries. users per second: 13672.4

recommendations finished on 89000/100638 queries. users per second: 13680.5

recommendations finished on 90000/100638 queries. users per second: 13676.7

recommendations finished on 91000/100638 queries. users per second: 13680.2

recommendations finished on 92000/100638 queries. users per second: 13685.8

recommendations finished on 93000/100638 queries. users per second: 13686

recommendations finished on 94000/100638 queries. users per second: 13692.4

recommendations finished on 95000/100638 queries. users per second: 13696

recommendations finished on 96000/100638 queries. users per second: 13703.4

recommendations finished on 97000/100638 queries. users per second: 13696.4

recommendations finished on 98000/100638 queries. users per second: 13681.6

recommendations finished on 99000/100638 queries. users per second: 13669.7

recommendations finished on 100000/100638 queries. users per second: 13628.6

+-------------------------------+-------------------------------+
|       customer_unique_id      |           product_id          |
+-------------------------------+-------------------------------+
| 4854e9b3feff728c13ee5fc7d1... | 36f60d45225e60c7da4558b070... |
| 4854e9b3feff728c13ee5fc7d1... | ca349c4d87c594b4995fcc4fa9... |
| 4854e9b3feff728c13ee5fc7d1... | c9058c144a2fef1c5b35f31945... |
| 4854e9b3feff728c13ee5fc7d1... | 0bcc3eeca39e1064258aa1e932... |
| 4854e9b3feff728c13ee5fc7d1... | 705b4a0072a3ecf575b497f895... |
| 830d5b7aaa3b6f1e9ad63703be... | 36f60d45225e60c7da4558b070... |
| 830d5b7aaa3b6f1e9ad63703be... | ca349c4d87c594b4995fcc4fa9... |
| 830d5b7aaa3b6f1e9ad63703be... | c9058c144a2fef1c5b35f31945... |
| 830d5b7aaa3b6f1e9ad63703be... | 0bcc3eeca39e1064258aa1e932... |
| 830d5b7aaa3b6f1e9ad63703be... | 705b4a0072a3ecf575b497f895... |
| 87776adb449c551e74c13fc34f... | 36f60d45225e60c7da4558b070... |
| 87776adb449c551e74c13fc34f... | ca349c4d87c594b4995fcc4fa9... |
| 87776adb

In [63]:
name="cosine"
target="review_score"
cosine_norm= model(train_data_norm, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 80510 observations with 76144 users and 28388 items.

Data prepared in: 0.121678s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 758us                          | 1.25       |

| 9.278ms                        | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.51s                               | 0                | 0               |

| 3.65s                               | 100              | 28388           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 3.6755s

recommendations finished on 1000/100638 queries. users per second: 13705.2

recommendations finished on 2000/100638 queries. users per second: 13680.8

recommendations finished on 3000/100638 queries. users per second: 13758.9

recommendations finished on 4000/100638 queries. users per second: 13763.1

recommendations finished on 5000/100638 queries. users per second: 13814.1

recommendations finished on 6000/100638 queries. users per second: 13774.2

recommendations finished on 7000/100638 queries. users per second: 13797

recommendations finished on 8000/100638 queries. users per second: 13833.7

recommendations finished on 9000/100638 queries. users per second: 13870.8

recommendations finished on 10000/100638 queries. users per second: 13776.1

recommendations finished on 11000/100638 queries. users per second: 13762.1

recommendations finished on 12000/100638 queries. users per second: 13690.3

recommendations finished on 13000/100638 queries. users per second: 13578.4

recommendations finished on 14000/100638 queries. users per second: 13589.8

recommendations finished on 15000/100638 queries. users per second: 13492.2

recommendations finished on 16000/100638 queries. users per second: 13448.2

recommendations finished on 17000/100638 queries. users per second: 13455.9

recommendations finished on 18000/100638 queries. users per second: 13393.5

recommendations finished on 19000/100638 queries. users per second: 13286

recommendations finished on 20000/100638 queries. users per second: 13217.3

recommendations finished on 21000/100638 queries. users per second: 13225.1

recommendations finished on 22000/100638 queries. users per second: 13188.1

recommendations finished on 23000/100638 queries. users per second: 13201.2

recommendations finished on 24000/100638 queries. users per second: 13197.3

recommendations finished on 25000/100638 queries. users per second: 13152

recommendations finished on 26000/100638 queries. users per second: 13110.1

recommendations finished on 27000/100638 queries. users per second: 13071.9

recommendations finished on 28000/100638 queries. users per second: 13038.3

recommendations finished on 29000/100638 queries. users per second: 13036.7

recommendations finished on 30000/100638 queries. users per second: 13012

recommendations finished on 31000/100638 queries. users per second: 12990.6

recommendations finished on 32000/100638 queries. users per second: 12979.9

recommendations finished on 33000/100638 queries. users per second: 12985.4

recommendations finished on 34000/100638 queries. users per second: 12963.8

recommendations finished on 35000/100638 queries. users per second: 12979.6

recommendations finished on 36000/100638 queries. users per second: 12943.3

recommendations finished on 37000/100638 queries. users per second: 12922.8

recommendations finished on 38000/100638 queries. users per second: 12880.5

recommendations finished on 39000/100638 queries. users per second: 12890.9

recommendations finished on 40000/100638 queries. users per second: 12899.6

recommendations finished on 41000/100638 queries. users per second: 12904.2

recommendations finished on 42000/100638 queries. users per second: 12929.3

recommendations finished on 43000/100638 queries. users per second: 12949.4

recommendations finished on 44000/100638 queries. users per second: 12972.5

recommendations finished on 45000/100638 queries. users per second: 12993.9

recommendations finished on 46000/100638 queries. users per second: 13017.7

recommendations finished on 47000/100638 queries. users per second: 13038.4

recommendations finished on 48000/100638 queries. users per second: 13045.3

recommendations finished on 49000/100638 queries. users per second: 13042

recommendations finished on 50000/100638 queries. users per second: 13061.9

recommendations finished on 51000/100638 queries. users per second: 13083

recommendations finished on 52000/100638 queries. users per second: 13102

recommendations finished on 53000/100638 queries. users per second: 13123.7

recommendations finished on 54000/100638 queries. users per second: 13134

recommendations finished on 55000/100638 queries. users per second: 13144.1

recommendations finished on 56000/100638 queries. users per second: 13156.1

recommendations finished on 57000/100638 queries. users per second: 13167.1

recommendations finished on 58000/100638 queries. users per second: 13170.8

recommendations finished on 59000/100638 queries. users per second: 13195

recommendations finished on 60000/100638 queries. users per second: 13204.4

recommendations finished on 61000/100638 queries. users per second: 13216

recommendations finished on 62000/100638 queries. users per second: 13228.9

recommendations finished on 63000/100638 queries. users per second: 13243.2

recommendations finished on 64000/100638 queries. users per second: 13257.3

recommendations finished on 65000/100638 queries. users per second: 13271

recommendations finished on 66000/100638 queries. users per second: 13283.8

recommendations finished on 67000/100638 queries. users per second: 13297

recommendations finished on 68000/100638 queries. users per second: 13307.2

recommendations finished on 69000/100638 queries. users per second: 13318.2

recommendations finished on 70000/100638 queries. users per second: 13328

recommendations finished on 71000/100638 queries. users per second: 13316.3

recommendations finished on 72000/100638 queries. users per second: 13326.6

recommendations finished on 73000/100638 queries. users per second: 13335.1

recommendations finished on 74000/100638 queries. users per second: 13345.4

recommendations finished on 75000/100638 queries. users per second: 13355.8

recommendations finished on 76000/100638 queries. users per second: 13363.9

recommendations finished on 77000/100638 queries. users per second: 13371.6

recommendations finished on 78000/100638 queries. users per second: 13373.6

recommendations finished on 79000/100638 queries. users per second: 13382.5

recommendations finished on 80000/100638 queries. users per second: 13390.3

recommendations finished on 81000/100638 queries. users per second: 13399

recommendations finished on 82000/100638 queries. users per second: 13408.5

recommendations finished on 83000/100638 queries. users per second: 13412.5

recommendations finished on 84000/100638 queries. users per second: 13421.1

recommendations finished on 85000/100638 queries. users per second: 13429.6

recommendations finished on 86000/100638 queries. users per second: 13438.6

recommendations finished on 87000/100638 queries. users per second: 13447.1

recommendations finished on 88000/100638 queries. users per second: 13456

recommendations finished on 89000/100638 queries. users per second: 13463.3

recommendations finished on 90000/100638 queries. users per second: 13471.8

recommendations finished on 91000/100638 queries. users per second: 13477.2

recommendations finished on 92000/100638 queries. users per second: 13475.7

recommendations finished on 93000/100638 queries. users per second: 13481.7

recommendations finished on 94000/100638 queries. users per second: 13486.7

recommendations finished on 95000/100638 queries. users per second: 13491.6

recommendations finished on 96000/100638 queries. users per second: 13493.1

recommendations finished on 97000/100638 queries. users per second: 13496.6

recommendations finished on 98000/100638 queries. users per second: 13496.6

recommendations finished on 99000/100638 queries. users per second: 13498.2

recommendations finished on 100000/100638 queries. users per second: 13468.8

+-------------------------------+-------------------------------+
|       customer_unique_id      |           product_id          |
+-------------------------------+-------------------------------+
| 4854e9b3feff728c13ee5fc7d1... | 0b33e28a934b68a0475650499b... |
| 4854e9b3feff728c13ee5fc7d1... | efe572a7ee9cf0317bbcd4ec8b... |
| 4854e9b3feff728c13ee5fc7d1... | e4b0f340d2552f744b4c2d48df... |
| 4854e9b3feff728c13ee5fc7d1... | 4fbee589d3377144a90a338e4a... |
| 4854e9b3feff728c13ee5fc7d1... | 57c806e92eec9168ff2065cc65... |
| 830d5b7aaa3b6f1e9ad63703be... | 0b33e28a934b68a0475650499b... |
| 830d5b7aaa3b6f1e9ad63703be... | efe572a7ee9cf0317bbcd4ec8b... |
| 830d5b7aaa3b6f1e9ad63703be... | e4b0f340d2552f744b4c2d48df... |
| 830d5b7aaa3b6f1e9ad63703be... | 4fbee589d3377144a90a338e4a... |
| 830d5b7aaa3b6f1e9ad63703be... | 57c806e92eec9168ff2065cc65... |
| 87776adb449c551e74c13fc34f... | 0b33e28a934b68a0475650499b... |
| 87776adb449c551e74c13fc34f... | efe572a7ee9cf0317bbcd4ec8b... |
| 87776adb

In [62]:
name="pearson"
target="review_score"
pearson= model(train_data, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 80510 observations with 76129 users and 28332 items.

Data prepared in: 0.138333s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 11.683ms                       | 1.25       |

| 20.838ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.51s                               | 0                | 0               |

| 3.51s                               | 100              | 28332           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 3.54646s

recommendations finished on 1000/100638 queries. users per second: 13882.5

recommendations finished on 2000/100638 queries. users per second: 13538.8

recommendations finished on 3000/100638 queries. users per second: 13427.1

recommendations finished on 4000/100638 queries. users per second: 13389.5

recommendations finished on 5000/100638 queries. users per second: 13360.2

recommendations finished on 6000/100638 queries. users per second: 13330.6

recommendations finished on 7000/100638 queries. users per second: 13327.2

recommendations finished on 8000/100638 queries. users per second: 13295.8

recommendations finished on 9000/100638 queries. users per second: 13297.4

recommendations finished on 10000/100638 queries. users per second: 13289.5

recommendations finished on 11000/100638 queries. users per second: 13265.5

recommendations finished on 12000/100638 queries. users per second: 13267.5

recommendations finished on 13000/100638 queries. users per second: 13257.1

recommendations finished on 14000/100638 queries. users per second: 13257.1

recommendations finished on 15000/100638 queries. users per second: 13254.4

recommendations finished on 16000/100638 queries. users per second: 13242.6

recommendations finished on 17000/100638 queries. users per second: 13238

recommendations finished on 18000/100638 queries. users per second: 13224.7

recommendations finished on 19000/100638 queries. users per second: 13188.5

recommendations finished on 20000/100638 queries. users per second: 13141.3

recommendations finished on 21000/100638 queries. users per second: 13117.5

recommendations finished on 22000/100638 queries. users per second: 13110.8

recommendations finished on 23000/100638 queries. users per second: 13113.4

recommendations finished on 24000/100638 queries. users per second: 13103

recommendations finished on 25000/100638 queries. users per second: 13092.2

recommendations finished on 26000/100638 queries. users per second: 13081.3

recommendations finished on 27000/100638 queries. users per second: 13074.1

recommendations finished on 28000/100638 queries. users per second: 13073.6

recommendations finished on 29000/100638 queries. users per second: 13074.6

recommendations finished on 30000/100638 queries. users per second: 13068.6

recommendations finished on 31000/100638 queries. users per second: 13070.1

recommendations finished on 32000/100638 queries. users per second: 13074.4

recommendations finished on 33000/100638 queries. users per second: 13074.5

recommendations finished on 34000/100638 queries. users per second: 13063

recommendations finished on 35000/100638 queries. users per second: 13079.5

recommendations finished on 36000/100638 queries. users per second: 13099.9

recommendations finished on 37000/100638 queries. users per second: 13072.1

recommendations finished on 38000/100638 queries. users per second: 13082.5

recommendations finished on 39000/100638 queries. users per second: 13065.7

recommendations finished on 40000/100638 queries. users per second: 13077.1

recommendations finished on 41000/100638 queries. users per second: 13084.8

recommendations finished on 42000/100638 queries. users per second: 13095.7

recommendations finished on 43000/100638 queries. users per second: 13104.6

recommendations finished on 44000/100638 queries. users per second: 13110.4

recommendations finished on 45000/100638 queries. users per second: 13118

recommendations finished on 46000/100638 queries. users per second: 13124.4

recommendations finished on 47000/100638 queries. users per second: 13130

recommendations finished on 48000/100638 queries. users per second: 13137.8

recommendations finished on 49000/100638 queries. users per second: 13141.7

recommendations finished on 50000/100638 queries. users per second: 13144.9

recommendations finished on 51000/100638 queries. users per second: 13144.9

recommendations finished on 52000/100638 queries. users per second: 13148

recommendations finished on 53000/100638 queries. users per second: 13152.5

recommendations finished on 54000/100638 queries. users per second: 13159.3

recommendations finished on 55000/100638 queries. users per second: 13164.6

recommendations finished on 56000/100638 queries. users per second: 13168.5

recommendations finished on 57000/100638 queries. users per second: 13174.3

recommendations finished on 58000/100638 queries. users per second: 13178.7

recommendations finished on 59000/100638 queries. users per second: 13183.1

recommendations finished on 60000/100638 queries. users per second: 13185.6

recommendations finished on 61000/100638 queries. users per second: 13186.6

recommendations finished on 62000/100638 queries. users per second: 13191.8

recommendations finished on 63000/100638 queries. users per second: 13193.6

recommendations finished on 64000/100638 queries. users per second: 13191.7

recommendations finished on 65000/100638 queries. users per second: 13190.9

recommendations finished on 66000/100638 queries. users per second: 13190.3

recommendations finished on 67000/100638 queries. users per second: 13195.1

recommendations finished on 68000/100638 queries. users per second: 13199.7

recommendations finished on 69000/100638 queries. users per second: 13204.4

recommendations finished on 70000/100638 queries. users per second: 13208.4

recommendations finished on 71000/100638 queries. users per second: 13212

recommendations finished on 72000/100638 queries. users per second: 13216.7

recommendations finished on 73000/100638 queries. users per second: 13220.7

recommendations finished on 74000/100638 queries. users per second: 13224.6

recommendations finished on 75000/100638 queries. users per second: 13226.3

recommendations finished on 76000/100638 queries. users per second: 13230.1

recommendations finished on 77000/100638 queries. users per second: 13229

recommendations finished on 78000/100638 queries. users per second: 13233.7

recommendations finished on 79000/100638 queries. users per second: 13238.1

recommendations finished on 80000/100638 queries. users per second: 13241.3

recommendations finished on 81000/100638 queries. users per second: 13244.7

recommendations finished on 82000/100638 queries. users per second: 13247.1

recommendations finished on 83000/100638 queries. users per second: 13248.5

recommendations finished on 84000/100638 queries. users per second: 13241.3

recommendations finished on 85000/100638 queries. users per second: 13244.1

recommendations finished on 86000/100638 queries. users per second: 13247

recommendations finished on 87000/100638 queries. users per second: 13251.1

recommendations finished on 88000/100638 queries. users per second: 13253.2

recommendations finished on 89000/100638 queries. users per second: 13258.4

recommendations finished on 90000/100638 queries. users per second: 13261.2

recommendations finished on 91000/100638 queries. users per second: 13262.6

recommendations finished on 92000/100638 queries. users per second: 13263.4

recommendations finished on 93000/100638 queries. users per second: 13265.5

recommendations finished on 94000/100638 queries. users per second: 13268.4

recommendations finished on 95000/100638 queries. users per second: 13268.2

recommendations finished on 96000/100638 queries. users per second: 13269.4

recommendations finished on 97000/100638 queries. users per second: 13268.1

recommendations finished on 98000/100638 queries. users per second: 13260.9

recommendations finished on 99000/100638 queries. users per second: 13190

recommendations finished on 100000/100638 queries. users per second: 13048.9

+-------------------------------+-------------------------------+-------+------+
|       customer_unique_id      |           product_id          | score | rank |
+-------------------------------+-------------------------------+-------+------+
| 4854e9b3feff728c13ee5fc7d1... | 6ed93af03d1f53308d3a9c6555... |  5.0  |  1   |
| 4854e9b3feff728c13ee5fc7d1... | f4851ffef385d506603396041d... |  5.0  |  2   |
| 4854e9b3feff728c13ee5fc7d1... | 18047039004204f52054f74b95... |  5.0  |  3   |
| 4854e9b3feff728c13ee5fc7d1... | 97868e08c02c9f4f6b49ebe5a4... |  5.0  |  4   |
| 4854e9b3feff728c13ee5fc7d1... | 73abf6f6dfa98d1d04760f4cc1... |  5.0  |  5   |
| 830d5b7aaa3b6f1e9ad63703be... | 6ed93af03d1f53308d3a9c6555... |  5.0  |  1   |
| 830d5b7aaa3b6f1e9ad63703be... | f4851ffef385d506603396041d... |  5.0  |  2   |
| 830d5b7aaa3b6f1e9ad63703be... | 18047039004204f52054f74b95... |  5.0  |  3   |
| 830d5b7aaa3b6f1e9ad63703be... | 97868e08c02c9f4f6b49ebe5a4... |  5.0  |  4   |
| 830d5b7aaa3b6f1e9ad63703be

In [64]:
name="pearson"
target="review_score"
pearson_norm= model(train_data_norm, name, user_id, item_id, target, users_to_recommend, n_rec, n_display)

Preparing data set.

Data has 80510 observations with 76144 users and 28388 items.

Data prepared in: 0.124896s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 1.261ms                        | 1.25       |

| 12.458ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.49s                               | 0                | 0               |

| 3.38s                               | 100              | 28388           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 3.40028s

recommendations finished on 1000/100638 queries. users per second: 12710.4

recommendations finished on 2000/100638 queries. users per second: 12905.8

recommendations finished on 3000/100638 queries. users per second: 12967.4

recommendations finished on 4000/100638 queries. users per second: 13010.7

recommendations finished on 5000/100638 queries. users per second: 13037.7

recommendations finished on 6000/100638 queries. users per second: 13077

recommendations finished on 7000/100638 queries. users per second: 13102.5

recommendations finished on 8000/100638 queries. users per second: 13111.9

recommendations finished on 9000/100638 queries. users per second: 13153.3

recommendations finished on 10000/100638 queries. users per second: 13185.5

recommendations finished on 11000/100638 queries. users per second: 13184.1

recommendations finished on 12000/100638 queries. users per second: 13208.7

recommendations finished on 13000/100638 queries. users per second: 13212.9

recommendations finished on 14000/100638 queries. users per second: 13230.3

recommendations finished on 15000/100638 queries. users per second: 13216.7

recommendations finished on 16000/100638 queries. users per second: 13181.3

recommendations finished on 17000/100638 queries. users per second: 13172.5

recommendations finished on 18000/100638 queries. users per second: 13166.2

recommendations finished on 19000/100638 queries. users per second: 13146.7

recommendations finished on 20000/100638 queries. users per second: 13134.6

recommendations finished on 21000/100638 queries. users per second: 13132.5

recommendations finished on 22000/100638 queries. users per second: 13126

recommendations finished on 23000/100638 queries. users per second: 13124.2

recommendations finished on 24000/100638 queries. users per second: 13130

recommendations finished on 25000/100638 queries. users per second: 13114.5

recommendations finished on 26000/100638 queries. users per second: 13108.5

recommendations finished on 27000/100638 queries. users per second: 13107.9

recommendations finished on 28000/100638 queries. users per second: 13106.2

recommendations finished on 29000/100638 queries. users per second: 13114

recommendations finished on 30000/100638 queries. users per second: 13114

recommendations finished on 31000/100638 queries. users per second: 13116.5

recommendations finished on 32000/100638 queries. users per second: 13108.7

recommendations finished on 33000/100638 queries. users per second: 13112.6

recommendations finished on 34000/100638 queries. users per second: 13096.5

recommendations finished on 35000/100638 queries. users per second: 13114.6

recommendations finished on 36000/100638 queries. users per second: 13126

recommendations finished on 37000/100638 queries. users per second: 13116.1

recommendations finished on 38000/100638 queries. users per second: 13137.1

recommendations finished on 39000/100638 queries. users per second: 13152.5

recommendations finished on 40000/100638 queries. users per second: 13175.8

recommendations finished on 41000/100638 queries. users per second: 13195.9

recommendations finished on 42000/100638 queries. users per second: 13216.7

recommendations finished on 43000/100638 queries. users per second: 13232.5

recommendations finished on 44000/100638 queries. users per second: 13250

recommendations finished on 45000/100638 queries. users per second: 13266.5

recommendations finished on 46000/100638 queries. users per second: 13280.3

recommendations finished on 47000/100638 queries. users per second: 13291.4

recommendations finished on 48000/100638 queries. users per second: 13308.2

recommendations finished on 49000/100638 queries. users per second: 13324.1

recommendations finished on 50000/100638 queries. users per second: 13336.3

recommendations finished on 51000/100638 queries. users per second: 13346.6

recommendations finished on 52000/100638 queries. users per second: 13355.7

recommendations finished on 53000/100638 queries. users per second: 13361.8

recommendations finished on 54000/100638 queries. users per second: 13373.4

recommendations finished on 55000/100638 queries. users per second: 13381.1

recommendations finished on 56000/100638 queries. users per second: 13391

recommendations finished on 57000/100638 queries. users per second: 13401

recommendations finished on 58000/100638 queries. users per second: 13412.2

recommendations finished on 59000/100638 queries. users per second: 13420

recommendations finished on 60000/100638 queries. users per second: 13429.3

recommendations finished on 61000/100638 queries. users per second: 13438

recommendations finished on 62000/100638 queries. users per second: 13444

recommendations finished on 63000/100638 queries. users per second: 13452.7

recommendations finished on 64000/100638 queries. users per second: 13460.5

recommendations finished on 65000/100638 queries. users per second: 13468.7

recommendations finished on 66000/100638 queries. users per second: 13478.8

recommendations finished on 67000/100638 queries. users per second: 13485.4

recommendations finished on 68000/100638 queries. users per second: 13491.8

recommendations finished on 69000/100638 queries. users per second: 13494.7

recommendations finished on 70000/100638 queries. users per second: 13502.7

recommendations finished on 71000/100638 queries. users per second: 13508.8

recommendations finished on 72000/100638 queries. users per second: 13510.8

recommendations finished on 73000/100638 queries. users per second: 13508.2

recommendations finished on 74000/100638 queries. users per second: 13514.8

recommendations finished on 75000/100638 queries. users per second: 13516.2

recommendations finished on 76000/100638 queries. users per second: 13520.5

recommendations finished on 77000/100638 queries. users per second: 13527.5

recommendations finished on 78000/100638 queries. users per second: 13531.6

recommendations finished on 79000/100638 queries. users per second: 13538.4

recommendations finished on 80000/100638 queries. users per second: 13545

recommendations finished on 81000/100638 queries. users per second: 13547.4

recommendations finished on 82000/100638 queries. users per second: 13551.2

recommendations finished on 83000/100638 queries. users per second: 13557.7

recommendations finished on 84000/100638 queries. users per second: 13561.6

recommendations finished on 85000/100638 queries. users per second: 13568.4

recommendations finished on 86000/100638 queries. users per second: 13575.6

recommendations finished on 87000/100638 queries. users per second: 13581.1

recommendations finished on 88000/100638 queries. users per second: 13583.9

recommendations finished on 89000/100638 queries. users per second: 13577

recommendations finished on 90000/100638 queries. users per second: 13578.7

recommendations finished on 91000/100638 queries. users per second: 13582.9

recommendations finished on 92000/100638 queries. users per second: 13587.3

recommendations finished on 93000/100638 queries. users per second: 13592

recommendations finished on 94000/100638 queries. users per second: 13595

recommendations finished on 95000/100638 queries. users per second: 13598.6

recommendations finished on 96000/100638 queries. users per second: 13596.5

recommendations finished on 97000/100638 queries. users per second: 13601.2

recommendations finished on 98000/100638 queries. users per second: 13606.1

recommendations finished on 99000/100638 queries. users per second: 13607.1

recommendations finished on 100000/100638 queries. users per second: 13597

+-------------------------------+-------------------------------+
|       customer_unique_id      |           product_id          |
+-------------------------------+-------------------------------+
| 4854e9b3feff728c13ee5fc7d1... | 5ca6a40fbc611451486b6525c6... |
| 4854e9b3feff728c13ee5fc7d1... | 50571d0f0f8ffbeec1c64cefee... |
| 4854e9b3feff728c13ee5fc7d1... | 1e1172a8b6ae12312c22ccf336... |
| 4854e9b3feff728c13ee5fc7d1... | cc711d05cc2a7eaabf0629dbda... |
| 4854e9b3feff728c13ee5fc7d1... | 1ba7948f3aa31cdaf78a4885f3... |
| 830d5b7aaa3b6f1e9ad63703be... | 5ca6a40fbc611451486b6525c6... |
| 830d5b7aaa3b6f1e9ad63703be... | 50571d0f0f8ffbeec1c64cefee... |
| 830d5b7aaa3b6f1e9ad63703be... | 1e1172a8b6ae12312c22ccf336... |
| 830d5b7aaa3b6f1e9ad63703be... | cc711d05cc2a7eaabf0629dbda... |
| 830d5b7aaa3b6f1e9ad63703be... | 1ba7948f3aa31cdaf78a4885f3... |
| 87776adb449c551e74c13fc34f... | 5ca6a40fbc611451486b6525c6... |
| 87776adb449c551e74c13fc34f... | 50571d0f0f8ffbeec1c64cefee... |
| 87776adb

## Evaluation on the test set


In [65]:
models_w_counts = [popularity, cosine, pearson]
models_w_norm = [pop_norm, cosine_norm, pearson_norm]
names_w_counts = ['Popularity Model on Review_score', 'Cosine Similarity on Review Score', 'Pearson Similarity on Review Score']
names_w_norm = ['Popularity Model on Scaled Reviews', 'Cosine Similarity on Scaled Review', 'Pearson Similarity on Scaled Review']

In [66]:
eval_counts = tc.recommender.util.compare_models(test_data, models_w_counts, model_names=names_w_counts)

eval_norm = tc.recommender.util.compare_models(test_data_norm, models_w_norm, model_names=names_w_norm)

PROGRESS: Evaluate model Popularity Model on Review_score


recommendations finished on 1000/19807 queries. users per second: 12689.2

recommendations finished on 2000/19807 queries. users per second: 11710.6

recommendations finished on 3000/19807 queries. users per second: 11620.3

recommendations finished on 4000/19807 queries. users per second: 11695.4

recommendations finished on 5000/19807 queries. users per second: 11743.5

recommendations finished on 6000/19807 queries. users per second: 11835.9

recommendations finished on 7000/19807 queries. users per second: 11880.1

recommendations finished on 8000/19807 queries. users per second: 11940.4

recommendations finished on 9000/19807 queries. users per second: 11992.3

recommendations finished on 10000/19807 queries. users per second: 11914.1

recommendations finished on 11000/19807 queries. users per second: 11934.3

recommendations finished on 12000/19807 queries. users per second: 11926.7

recommendations finished on 13000/19807 queries. users per second: 11920.4

recommendations finished on 14000/19807 queries. users per second: 11854.9

recommendations finished on 15000/19807 queries. users per second: 11872.2

recommendations finished on 16000/19807 queries. users per second: 11893.7

recommendations finished on 17000/19807 queries. users per second: 11896.5

recommendations finished on 18000/19807 queries. users per second: 11890.7

recommendations finished on 19000/19807 queries. users per second: 11826.4


Precision and recall summary statistics by cutoff
+--------+------------------------+-----------------------+
| cutoff |     mean_precision     |      mean_recall      |
+--------+------------------------+-----------------------+
|   1    |          0.0           |          0.0          |
|   2    |          0.0           |          0.0          |
|   3    | 1.6829067164807066e-05 | 5.048720149442118e-05 |
|   4    | 1.2621800373605294e-05 | 5.048720149442118e-05 |
|   5    | 1.0097440298884248e-05 | 5.048720149442118e-05 |
|   6    | 8.414533582403511e-06  | 5.048720149442118e-05 |
|   7    | 7.212457356345875e-06  | 5.048720149442118e-05 |
|   8    | 6.310900186802647e-06  | 5.048720149442118e-05 |
|   9    | 5.6096890549356836e-06 | 5.048720149442118e-05 |
|   10   | 5.048720149442139e-06  | 5.048720149442118e-05 |
+--------+------------------------+-----------------------+
[10 rows x 3 columns]


Overall RMSE: 1.4900980518250058

Per User RMSE (best)
+-----------------------------

recommendations finished on 1000/19807 queries. users per second: 13437.1

recommendations finished on 2000/19807 queries. users per second: 12659.8

recommendations finished on 3000/19807 queries. users per second: 12699.5

recommendations finished on 4000/19807 queries. users per second: 12800.4

recommendations finished on 5000/19807 queries. users per second: 12823.4

recommendations finished on 6000/19807 queries. users per second: 12786.1

recommendations finished on 7000/19807 queries. users per second: 12754.9

recommendations finished on 8000/19807 queries. users per second: 12778.5

recommendations finished on 9000/19807 queries. users per second: 12866.7

recommendations finished on 10000/19807 queries. users per second: 12965.9

recommendations finished on 11000/19807 queries. users per second: 13016.5

recommendations finished on 12000/19807 queries. users per second: 13031.8

recommendations finished on 13000/19807 queries. users per second: 13062

recommendations finished on 14000/19807 queries. users per second: 13078.8

recommendations finished on 15000/19807 queries. users per second: 13125.1

recommendations finished on 16000/19807 queries. users per second: 13165.2

recommendations finished on 17000/19807 queries. users per second: 13183

recommendations finished on 18000/19807 queries. users per second: 13206.8

recommendations finished on 19000/19807 queries. users per second: 13223.1


Precision and recall summary statistics by cutoff
+--------+-----------------------+----------------------+
| cutoff |     mean_precision    |     mean_recall      |
+--------+-----------------------+----------------------+
|   1    |  0.005452617761397492 | 0.005217010821090185 |
|   2    |  0.003660322108345536 | 0.007009306474142136 |
|   3    |  0.002675821679204326 | 0.007690883694316823 |
|   4    | 0.0023981420709850076 | 0.00923074333989666  |
|   5    | 0.0020295855000757295 | 0.009735615354840894 |
|   6    | 0.0017670520523047407 | 0.010164756567543455 |
|   7    |  0.00153625341690167  | 0.01031621817202673  |
|   8    | 0.0013757762407229797 | 0.010568654179498836 |
|   9    |  0.001234131592085849 | 0.010669628582487677 |
|   10   | 0.0011258645933255955 | 0.010821090186970932 |
+--------+-----------------------+----------------------+
[10 rows x 3 columns]


Overall RMSE: 4.282568995080432

Per User RMSE (best)
+-------------------------------+----------------------+---

recommendations finished on 1000/19807 queries. users per second: 11174.7

recommendations finished on 2000/19807 queries. users per second: 11010.4

recommendations finished on 3000/19807 queries. users per second: 11071.1

recommendations finished on 4000/19807 queries. users per second: 11147.1

recommendations finished on 5000/19807 queries. users per second: 11163.8

recommendations finished on 6000/19807 queries. users per second: 11165.5

recommendations finished on 7000/19807 queries. users per second: 11183.5

recommendations finished on 8000/19807 queries. users per second: 11190.2

recommendations finished on 9000/19807 queries. users per second: 11211.1

recommendations finished on 10000/19807 queries. users per second: 11207.1

recommendations finished on 11000/19807 queries. users per second: 11214.3

recommendations finished on 12000/19807 queries. users per second: 11205.5

recommendations finished on 13000/19807 queries. users per second: 11129.3

recommendations finished on 14000/19807 queries. users per second: 11107.2

recommendations finished on 15000/19807 queries. users per second: 11063.9

recommendations finished on 16000/19807 queries. users per second: 11072.8

recommendations finished on 17000/19807 queries. users per second: 11082.2

recommendations finished on 18000/19807 queries. users per second: 11098.3

recommendations finished on 19000/19807 queries. users per second: 11083


Precision and recall summary statistics by cutoff
+--------+------------------------+-----------------------+
| cutoff |     mean_precision     |      mean_recall      |
+--------+------------------------+-----------------------+
|   1    |          0.0           |          0.0          |
|   2    |          0.0           |          0.0          |
|   3    |          0.0           |          0.0          |
|   4    |          0.0           |          0.0          |
|   5    |          0.0           |          0.0          |
|   6    |          0.0           |          0.0          |
|   7    |          0.0           |          0.0          |
|   8    | 6.310900186802631e-06  | 5.048720149442105e-05 |
|   9    | 5.6096890549356785e-06 | 5.048720149442116e-05 |
|   10   | 5.048720149442119e-06  | 5.048720149442108e-05 |
+--------+------------------------+-----------------------+
[10 rows x 3 columns]


Overall RMSE: 2.428166957507819

Per User RMSE (best)
+------------------------------

recommendations finished on 1000/19838 queries. users per second: 13631

recommendations finished on 2000/19838 queries. users per second: 13514.8

recommendations finished on 3000/19838 queries. users per second: 13495.5

recommendations finished on 4000/19838 queries. users per second: 13613.7

recommendations finished on 5000/19838 queries. users per second: 13558.8

recommendations finished on 6000/19838 queries. users per second: 13617

recommendations finished on 7000/19838 queries. users per second: 13573.7

recommendations finished on 8000/19838 queries. users per second: 13616.4

recommendations finished on 9000/19838 queries. users per second: 13593.8

recommendations finished on 10000/19838 queries. users per second: 13635.6

recommendations finished on 11000/19838 queries. users per second: 13599.2

recommendations finished on 12000/19838 queries. users per second: 13586.8

recommendations finished on 13000/19838 queries. users per second: 13493.1

recommendations finished on 14000/19838 queries. users per second: 13482

recommendations finished on 15000/19838 queries. users per second: 13468.6

recommendations finished on 16000/19838 queries. users per second: 13486

recommendations finished on 17000/19838 queries. users per second: 13470.8

recommendations finished on 18000/19838 queries. users per second: 13464.2

recommendations finished on 19000/19838 queries. users per second: 13435.8


Precision and recall summary statistics by cutoff
+--------+------------------------+-----------------------+
| cutoff |     mean_precision     |      mean_recall      |
+--------+------------------------+-----------------------+
|   1    | 0.0001008166145780825  | 0.0001008166145780825 |
|   2    | 0.0001008166145780825  |  0.000201633229156165 |
|   3    | 0.0001344221527707764  |  0.00040326645831233  |
|   4    | 0.00011341869140034283 | 0.0004536747656013713 |
|   5    | 0.00011089827603589066 | 0.0005544913801794518 |
|   6    | 0.00010081661457808254 | 0.0006048996874684941 |
|   7    | 0.00012242017484481434 | 0.0008569412239137011 |
|   8    | 0.00013232180663373327 | 0.0010585744530698662 |
|   9    | 0.00011761938367442962 | 0.0010585744530698655 |
|   10   | 0.00011593910676479494 | 0.0011593910676479494 |
+--------+------------------------+-----------------------+
[10 rows x 3 columns]


Overall RMSE: 0.27494936913444645

Per User RMSE (best)
+----------------------------

recommendations finished on 1000/19838 queries. users per second: 13455.9

recommendations finished on 2000/19838 queries. users per second: 13059.1

recommendations finished on 3000/19838 queries. users per second: 13158.3

recommendations finished on 4000/19838 queries. users per second: 13312.7

recommendations finished on 5000/19838 queries. users per second: 13264.4

recommendations finished on 6000/19838 queries. users per second: 13257.3

recommendations finished on 7000/19838 queries. users per second: 13272.8

recommendations finished on 8000/19838 queries. users per second: 13351

recommendations finished on 9000/19838 queries. users per second: 13359.4

recommendations finished on 10000/19838 queries. users per second: 13357.3

recommendations finished on 11000/19838 queries. users per second: 13384.8

recommendations finished on 12000/19838 queries. users per second: 13367.6

recommendations finished on 13000/19838 queries. users per second: 13277.1

recommendations finished on 14000/19838 queries. users per second: 13220.7

recommendations finished on 15000/19838 queries. users per second: 13234.5

recommendations finished on 16000/19838 queries. users per second: 13248.2

recommendations finished on 17000/19838 queries. users per second: 13226.3

recommendations finished on 18000/19838 queries. users per second: 13223.6

recommendations finished on 19000/19838 queries. users per second: 13216.1


Precision and recall summary statistics by cutoff
+--------+------------------------+-----------------------+
| cutoff |     mean_precision     |      mean_recall      |
+--------+------------------------+-----------------------+
|   1    | 0.0012602076822260339  |  0.001209799374936991 |
|   2    | 0.0010081661457808269  | 0.0019407198306280917 |
|   3    | 0.0008569412239137013  |  0.002470007057163023 |
|   4    | 0.0006931142252243168  |  0.002671640286319185 |
|   5    | 0.0005847363645528776  |  0.002822865208186312 |
|   6    | 0.0004956816883422395  | 0.0028732735154753533 |
|   7    | 0.0004320712053346394  | 0.0029236818227643914 |
|   8    | 0.0003780623046678098  |  0.002923681822764396 |
|   9    | 0.0003528581510232886  |  0.003074906744631518 |
|   10   | 0.00033269482810767165 |  0.003226131666498644 |
+--------+------------------------+-----------------------+
[10 rows x 3 columns]


Overall RMSE: 0.2484541891668454

Per User RMSE (best)
+-----------------------------

recommendations finished on 1000/19838 queries. users per second: 12638.6

recommendations finished on 2000/19838 queries. users per second: 12146.4

recommendations finished on 3000/19838 queries. users per second: 12187.2

recommendations finished on 4000/19838 queries. users per second: 12291.4

recommendations finished on 5000/19838 queries. users per second: 12254

recommendations finished on 6000/19838 queries. users per second: 12237.9

recommendations finished on 7000/19838 queries. users per second: 12230.6

recommendations finished on 8000/19838 queries. users per second: 12226.8

recommendations finished on 9000/19838 queries. users per second: 12281.8

recommendations finished on 10000/19838 queries. users per second: 12265.2

recommendations finished on 11000/19838 queries. users per second: 12236.7

recommendations finished on 12000/19838 queries. users per second: 12200.5

recommendations finished on 13000/19838 queries. users per second: 12175.5

recommendations finished on 14000/19838 queries. users per second: 12147.8

recommendations finished on 15000/19838 queries. users per second: 12160.4

recommendations finished on 16000/19838 queries. users per second: 12162.1

recommendations finished on 17000/19838 queries. users per second: 12097.1

recommendations finished on 18000/19838 queries. users per second: 12091.1

recommendations finished on 19000/19838 queries. users per second: 12075.2


Precision and recall summary statistics by cutoff
+--------+------------------------+-----------------------+
| cutoff |     mean_precision     |      mean_recall      |
+--------+------------------------+-----------------------+
|   1    |          0.0           |          0.0          |
|   2    |          0.0           |          0.0          |
|   3    |          0.0           |          0.0          |
|   4    | 1.2602076822260312e-05 | 5.040830728904125e-05 |
|   5    | 1.0081661457808212e-05 | 5.040830728904125e-05 |
|   6    |  8.40138454817354e-06  | 5.040830728904125e-05 |
|   7    | 7.201186755577316e-06  | 5.040830728904125e-05 |
|   8    | 6.301038411130156e-06  | 5.040830728904125e-05 |
|   9    |  5.60092303211569e-06  | 5.040830728904125e-05 |
|   10   | 5.0408307289041125e-06 | 5.040830728904125e-05 |
+--------+------------------------+-----------------------+
[10 rows x 3 columns]


Overall RMSE: 0.2749796642774072

Per User RMSE (best)
+-----------------------------

**Original Dataset**:


**Popularity Model** *RMSE*: 1.4900
**Cosine Similarity** *RMSE*: 4.2825
**Pearson Correlation** *RMSE*:  2.4281

**Normalized Dataset**:


**Popularity Model** *RMSE*: 0.2749
**Cosine Similarity** *RMSE*: 0.24
**Pearson Correlation** *RMSE*: 0.27

The best model is the Collaborative filtering model with Cosine Similarity and normalized and standardized data


The final recommendation is printed based on the lower RMSE

In [67]:
final_model = tc.item_similarity_recommender.create(tc.SFrame(df_norm), 
                                            user_id=user_id, 
                                            item_id=item_id, 
                                            target='review_score', similarity_type='cosine')
recom = final_model.recommend(users=users_to_recommend, k=n_rec)

recom.print_rows(n_display)

Preparing data set.

Data has 100638 observations with 94007 users and 32541 items.

Data prepared in: 0.166214s

Training model from provided data.

Gathering per-item and per-user statistics.

+--------------------------------+------------+

| Elapsed Time (Item Statistics) | % Complete |

+--------------------------------+------------+

| 882us                          | 1          |

| 12.234ms                       | 100        |

+--------------------------------+------------+

Setting up lookup tables.

Processing data in one pass using dense lookup tables.

+-------------------------------------+------------------+-----------------+

| Elapsed Time (Constructing Lookups) | Total % Complete | Items Processed |

+-------------------------------------+------------------+-----------------+

| 1.94s                               | 0                | 0               |

| 5.05s                               | 100              | 32541           |

+-------------------------------------+------------------+-----------------+

Finalizing lookup tables.

Generating candidate set for working with new users.

Finished training in 5.06808s

recommendations finished on 1000/100638 queries. users per second: 12621.3

recommendations finished on 2000/100638 queries. users per second: 12385.9

recommendations finished on 3000/100638 queries. users per second: 12307.1

recommendations finished on 4000/100638 queries. users per second: 12257.2

recommendations finished on 5000/100638 queries. users per second: 12216.8

recommendations finished on 6000/100638 queries. users per second: 12202.9

recommendations finished on 7000/100638 queries. users per second: 12182.7

recommendations finished on 8000/100638 queries. users per second: 12173.7

recommendations finished on 9000/100638 queries. users per second: 12174.5

recommendations finished on 10000/100638 queries. users per second: 12169.3

recommendations finished on 11000/100638 queries. users per second: 12179.9

recommendations finished on 12000/100638 queries. users per second: 12185.5

recommendations finished on 13000/100638 queries. users per second: 12180.2

recommendations finished on 14000/100638 queries. users per second: 12168.4

recommendations finished on 15000/100638 queries. users per second: 12011.9

recommendations finished on 16000/100638 queries. users per second: 11736.9

recommendations finished on 17000/100638 queries. users per second: 11512.7

recommendations finished on 18000/100638 queries. users per second: 11409.9

recommendations finished on 19000/100638 queries. users per second: 11259.2

recommendations finished on 20000/100638 queries. users per second: 11161.7

recommendations finished on 21000/100638 queries. users per second: 11031.8

recommendations finished on 22000/100638 queries. users per second: 10944.9

recommendations finished on 23000/100638 queries. users per second: 10850.1

recommendations finished on 24000/100638 queries. users per second: 10793.7

recommendations finished on 25000/100638 queries. users per second: 10729.6

recommendations finished on 26000/100638 queries. users per second: 10649

recommendations finished on 27000/100638 queries. users per second: 10612.1

recommendations finished on 28000/100638 queries. users per second: 10607.2

recommendations finished on 29000/100638 queries. users per second: 10644.7

recommendations finished on 30000/100638 queries. users per second: 10677

recommendations finished on 31000/100638 queries. users per second: 10707.9

recommendations finished on 32000/100638 queries. users per second: 10738.1

recommendations finished on 33000/100638 queries. users per second: 10767

recommendations finished on 34000/100638 queries. users per second: 10785.7

recommendations finished on 35000/100638 queries. users per second: 10815.5

recommendations finished on 36000/100638 queries. users per second: 10818.3

recommendations finished on 37000/100638 queries. users per second: 10854.4

recommendations finished on 38000/100638 queries. users per second: 10890.4

recommendations finished on 39000/100638 queries. users per second: 10911.9

recommendations finished on 40000/100638 queries. users per second: 10944.5

recommendations finished on 41000/100638 queries. users per second: 10977.3

recommendations finished on 42000/100638 queries. users per second: 11007.4

recommendations finished on 43000/100638 queries. users per second: 11033.5

recommendations finished on 44000/100638 queries. users per second: 11025.9

recommendations finished on 45000/100638 queries. users per second: 11003.4

recommendations finished on 46000/100638 queries. users per second: 10981.5

recommendations finished on 47000/100638 queries. users per second: 10948.5

recommendations finished on 48000/100638 queries. users per second: 10908.2

recommendations finished on 49000/100638 queries. users per second: 10922.3

recommendations finished on 50000/100638 queries. users per second: 10950.1

recommendations finished on 51000/100638 queries. users per second: 10976.3

recommendations finished on 52000/100638 queries. users per second: 11003.6

recommendations finished on 53000/100638 queries. users per second: 11029.6

recommendations finished on 54000/100638 queries. users per second: 11054.9

recommendations finished on 55000/100638 queries. users per second: 11077.8

recommendations finished on 56000/100638 queries. users per second: 11098.8

recommendations finished on 57000/100638 queries. users per second: 11119.2

recommendations finished on 58000/100638 queries. users per second: 11135.4

recommendations finished on 59000/100638 queries. users per second: 11101.3

recommendations finished on 60000/100638 queries. users per second: 11069

recommendations finished on 61000/100638 queries. users per second: 11048.3

recommendations finished on 62000/100638 queries. users per second: 11035.7

recommendations finished on 63000/100638 queries. users per second: 11028.2

recommendations finished on 64000/100638 queries. users per second: 11011

recommendations finished on 65000/100638 queries. users per second: 10997.5

recommendations finished on 66000/100638 queries. users per second: 10988.3

recommendations finished on 67000/100638 queries. users per second: 10967.9

recommendations finished on 68000/100638 queries. users per second: 10952.9

recommendations finished on 69000/100638 queries. users per second: 10946.5

recommendations finished on 70000/100638 queries. users per second: 10930.2

recommendations finished on 71000/100638 queries. users per second: 10916.4

recommendations finished on 72000/100638 queries. users per second: 10936.2

recommendations finished on 73000/100638 queries. users per second: 10951.7

recommendations finished on 74000/100638 queries. users per second: 10967.3

recommendations finished on 75000/100638 queries. users per second: 10976.7

recommendations finished on 76000/100638 queries. users per second: 10948.9

recommendations finished on 77000/100638 queries. users per second: 10940.4

recommendations finished on 78000/100638 queries. users per second: 10924.6

recommendations finished on 79000/100638 queries. users per second: 10923.9

recommendations finished on 80000/100638 queries. users per second: 10927.8

recommendations finished on 81000/100638 queries. users per second: 10942.1

recommendations finished on 82000/100638 queries. users per second: 10958.1

recommendations finished on 83000/100638 queries. users per second: 10974.2

recommendations finished on 84000/100638 queries. users per second: 10987.1

recommendations finished on 85000/100638 queries. users per second: 10967.1

recommendations finished on 86000/100638 queries. users per second: 10944

recommendations finished on 87000/100638 queries. users per second: 10942.3

recommendations finished on 88000/100638 queries. users per second: 10955.6

recommendations finished on 89000/100638 queries. users per second: 10969.4

recommendations finished on 90000/100638 queries. users per second: 10981.2

recommendations finished on 91000/100638 queries. users per second: 10973.3

recommendations finished on 92000/100638 queries. users per second: 10957.7

recommendations finished on 93000/100638 queries. users per second: 10952.9

recommendations finished on 94000/100638 queries. users per second: 10961.9

recommendations finished on 95000/100638 queries. users per second: 10975

recommendations finished on 96000/100638 queries. users per second: 10985

recommendations finished on 97000/100638 queries. users per second: 10965.2

recommendations finished on 98000/100638 queries. users per second: 10946.4

recommendations finished on 99000/100638 queries. users per second: 10933.1

recommendations finished on 100000/100638 queries. users per second: 10875

+-------------------------------+-------------------------------+
|       customer_unique_id      |           product_id          |
+-------------------------------+-------------------------------+
| 4854e9b3feff728c13ee5fc7d1... | ce49d8be84e535ca2a5eb0e76c... |
| 4854e9b3feff728c13ee5fc7d1... | e4b0f340d2552f744b4c2d48df... |
| 4854e9b3feff728c13ee5fc7d1... | efe572a7ee9cf0317bbcd4ec8b... |
| 4854e9b3feff728c13ee5fc7d1... | 0b33e28a934b68a0475650499b... |
| 4854e9b3feff728c13ee5fc7d1... | 370c500401de58da2a25c8cab2... |
| 830d5b7aaa3b6f1e9ad63703be... | ce49d8be84e535ca2a5eb0e76c... |
| 830d5b7aaa3b6f1e9ad63703be... | e4b0f340d2552f744b4c2d48df... |
| 830d5b7aaa3b6f1e9ad63703be... | efe572a7ee9cf0317bbcd4ec8b... |
| 830d5b7aaa3b6f1e9ad63703be... | 0b33e28a934b68a0475650499b... |
| 830d5b7aaa3b6f1e9ad63703be... | 370c500401de58da2a25c8cab2... |
| 87776adb449c551e74c13fc34f... | ce49d8be84e535ca2a5eb0e76c... |
| 87776adb449c551e74c13fc34f... | e4b0f340d2552f744b4c2d48df... |
| 87776adb

The frame produced through Turicreate is now transformed into a Pandas Dataframe, showing each cosine similarity score and the rank for each user

In [68]:
df_rec = recom.to_dataframe()
print(df_rec.shape)
df_rec.head()

(503190, 4)


customer_unique_id                        product_id  \
0  4854e9b3feff728c13ee5fc7d1547e92  ce49d8be84e535ca2a5eb0e76c09c348   
1  4854e9b3feff728c13ee5fc7d1547e92  e4b0f340d2552f744b4c2d48dffbc0e2   
2  4854e9b3feff728c13ee5fc7d1547e92  efe572a7ee9cf0317bbcd4ec8b98a088   
3  4854e9b3feff728c13ee5fc7d1547e92  0b33e28a934b68a0475650499b9ae509   
4  4854e9b3feff728c13ee5fc7d1547e92  370c500401de58da2a25c8cab27527cc   

      score  rank  
0  0.002964     1  
1  0.002444     2  
2  0.002422     3  
3  0.002230     4  
4  0.002097     5

A function has been created to make the five recommendation in one row divided by the bar symbol (*|*) to have a clean dataset

In [84]:
def create_output(model, users_to_recommend, n_rec, print_csv=True):
    recomendation = model.recommend(users=users_to_recommend, k=n_rec)
    df_rec = recomendation.to_dataframe()
    df_rec['recommendedProducts'] = df_rec.groupby([user_id])[item_id] \
        .transform(lambda x: '|'.join(x.astype(str)))
    df_output = df_rec[['customer_unique_id', 'recommendedProducts']].drop_duplicates() \
        .sort_values('customer_unique_id').set_index('customer_unique_id')
    if print_csv:
        df_output.to_csv('recommendation.csv')
    return df_output

In [85]:
df_output = create_output(cosine_norm, users_to_recommend, n_rec, print_csv=True)
df_output.head()

recommendations finished on 1000/100638 queries. users per second: 12777.9

recommendations finished on 2000/100638 queries. users per second: 12433.6

recommendations finished on 3000/100638 queries. users per second: 12395.4

recommendations finished on 4000/100638 queries. users per second: 12362.9

recommendations finished on 5000/100638 queries. users per second: 12384.3

recommendations finished on 6000/100638 queries. users per second: 12447.7

recommendations finished on 7000/100638 queries. users per second: 12443.5

recommendations finished on 8000/100638 queries. users per second: 12471.9

recommendations finished on 9000/100638 queries. users per second: 12473.2

recommendations finished on 10000/100638 queries. users per second: 12485

recommendations finished on 11000/100638 queries. users per second: 12506.6

recommendations finished on 12000/100638 queries. users per second: 12511.8

recommendations finished on 13000/100638 queries. users per second: 12464.9

recommendations finished on 14000/100638 queries. users per second: 12418.7

recommendations finished on 15000/100638 queries. users per second: 12432.2

recommendations finished on 16000/100638 queries. users per second: 12442

recommendations finished on 17000/100638 queries. users per second: 12455.8

recommendations finished on 18000/100638 queries. users per second: 12476

recommendations finished on 19000/100638 queries. users per second: 12486.6

recommendations finished on 20000/100638 queries. users per second: 12490.5

recommendations finished on 21000/100638 queries. users per second: 12497.9

recommendations finished on 22000/100638 queries. users per second: 12514.6

recommendations finished on 23000/100638 queries. users per second: 12528.7

recommendations finished on 24000/100638 queries. users per second: 12532.1

recommendations finished on 25000/100638 queries. users per second: 12535.8

recommendations finished on 26000/100638 queries. users per second: 12541.4

recommendations finished on 27000/100638 queries. users per second: 12533.8

recommendations finished on 28000/100638 queries. users per second: 12534

recommendations finished on 29000/100638 queries. users per second: 12537.9

recommendations finished on 30000/100638 queries. users per second: 12531

recommendations finished on 31000/100638 queries. users per second: 12530.1

recommendations finished on 32000/100638 queries. users per second: 12528

recommendations finished on 33000/100638 queries. users per second: 12529.2

recommendations finished on 34000/100638 queries. users per second: 12506.2

recommendations finished on 35000/100638 queries. users per second: 12491.3

recommendations finished on 36000/100638 queries. users per second: 12471.1

recommendations finished on 37000/100638 queries. users per second: 12434.6

recommendations finished on 38000/100638 queries. users per second: 12376.3

recommendations finished on 39000/100638 queries. users per second: 12352.2

recommendations finished on 40000/100638 queries. users per second: 12343.9

recommendations finished on 41000/100638 queries. users per second: 12315

recommendations finished on 42000/100638 queries. users per second: 12325

recommendations finished on 43000/100638 queries. users per second: 12330.7

recommendations finished on 44000/100638 queries. users per second: 12337.5

recommendations finished on 45000/100638 queries. users per second: 12338.6

recommendations finished on 46000/100638 queries. users per second: 12305.3

recommendations finished on 47000/100638 queries. users per second: 12287.3

recommendations finished on 48000/100638 queries. users per second: 12262

recommendations finished on 49000/100638 queries. users per second: 12257.7

recommendations finished on 50000/100638 queries. users per second: 12243.5

recommendations finished on 51000/100638 queries. users per second: 12217.8

recommendations finished on 52000/100638 queries. users per second: 12201

recommendations finished on 53000/100638 queries. users per second: 12187.6

recommendations finished on 54000/100638 queries. users per second: 12166

recommendations finished on 55000/100638 queries. users per second: 12165.7

recommendations finished on 56000/100638 queries. users per second: 12149.9

recommendations finished on 57000/100638 queries. users per second: 12115.4

recommendations finished on 58000/100638 queries. users per second: 12131.2

recommendations finished on 59000/100638 queries. users per second: 12143.8

recommendations finished on 60000/100638 queries. users per second: 12142.5

recommendations finished on 61000/100638 queries. users per second: 12130.8

recommendations finished on 62000/100638 queries. users per second: 12115.7

recommendations finished on 63000/100638 queries. users per second: 12119.4

recommendations finished on 64000/100638 queries. users per second: 12117.8

recommendations finished on 65000/100638 queries. users per second: 12111.4

recommendations finished on 66000/100638 queries. users per second: 12103.6

recommendations finished on 67000/100638 queries. users per second: 12118.6

recommendations finished on 68000/100638 queries. users per second: 12134.2

recommendations finished on 69000/100638 queries. users per second: 12150.2

recommendations finished on 70000/100638 queries. users per second: 12165.6

recommendations finished on 71000/100638 queries. users per second: 12180.9

recommendations finished on 72000/100638 queries. users per second: 12193.5

recommendations finished on 73000/100638 queries. users per second: 12206.7

recommendations finished on 74000/100638 queries. users per second: 12221.7

recommendations finished on 75000/100638 queries. users per second: 12234.8

recommendations finished on 76000/100638 queries. users per second: 12247.9

recommendations finished on 77000/100638 queries. users per second: 12259.4

recommendations finished on 78000/100638 queries. users per second: 12271.2

recommendations finished on 79000/100638 queries. users per second: 12282.9

recommendations finished on 80000/100638 queries. users per second: 12294.8

recommendations finished on 81000/100638 queries. users per second: 12306.2

recommendations finished on 82000/100638 queries. users per second: 12318.6

recommendations finished on 83000/100638 queries. users per second: 12331.6

recommendations finished on 84000/100638 queries. users per second: 12340

recommendations finished on 85000/100638 queries. users per second: 12337

recommendations finished on 86000/100638 queries. users per second: 12347.3

recommendations finished on 87000/100638 queries. users per second: 12358

recommendations finished on 88000/100638 queries. users per second: 12367

recommendations finished on 89000/100638 queries. users per second: 12378.7

recommendations finished on 90000/100638 queries. users per second: 12382.4

recommendations finished on 91000/100638 queries. users per second: 12382.4

recommendations finished on 92000/100638 queries. users per second: 12381.3

recommendations finished on 93000/100638 queries. users per second: 12374.4

recommendations finished on 94000/100638 queries. users per second: 12369.4

recommendations finished on 95000/100638 queries. users per second: 12374.3

recommendations finished on 96000/100638 queries. users per second: 12369.6

recommendations finished on 97000/100638 queries. users per second: 12375.9

recommendations finished on 98000/100638 queries. users per second: 12380

recommendations finished on 99000/100638 queries. users per second: 12384.1

recommendations finished on 100000/100638 queries. users per second: 12366

(94007, 1)


recommendedProducts
customer_unique_id                                                                 
0000366f3b9a7992bf8c76cfdf3221e2  cfe6e9c01d0bbb5df9a75f0e3286baa9|aa3cf7c656b2a...
0000b849f77a49e4a4ce2b2a4ca5be3f  0b33e28a934b68a0475650499b9ae509|efe572a7ee9cf...
0000f46a3911fa3c0805444483337064  0b33e28a934b68a0475650499b9ae509|efe572a7ee9cf...
0000f6ccb0745a6a4b88665a16c9f078  0b33e28a934b68a0475650499b9ae509|efe572a7ee9cf...
0004aac84e0df4da2b147fca70cf8255  0b33e28a934b68a0475650499b9ae509|efe572a7ee9cf...

Next step was to create a function that on user input tells which products are recommended to a specific user and if the user is not present in the database it return **"customer not found"**

In [2]:
def customer_recomendation():
    if customer_id not in df_output.index:
        print('Customer not found.')
        return customer_id
    return df_output.loc[customer_id]

In [87]:
customer_recomendation("44b6bbfea26596437062a38c8e6bcec1")

recommendedProducts    cfe6e9c01d0bbb5df9a75f0e3286baa9|aa3cf7c656b2a...
Name: 44b6bbfea26596437062a38c8e6bcec1, dtype: object

In [88]:
customer_recomendation("miro")

Customer not found.


'miro'